### PR TITLE
Adding Percona support

### PR DESCRIPTION
**Pull request by robison**
I've taken the liberty of adding Percona support to the `mysql` module, based on the documentation found at https://www.percona.com/doc/percona-server/5.6/diagnostics/slow_extended.html. I've tested it against a number of cases in our environment and it seems to work well, while not impacting the existing functionality.

### CODE REVIEW

**Comment by robison** [comment]
Added fully working tests for the Percona format.

**Comment by robison** [comment]
Also, I've updated it to better support hostname, address, and schema string formats. Feedback appreciated.

File: modules/mysql.lua [hunk 1]
```diff
@@ -6,35 +6,44 @@
...
 local query         = query_time * space * lock_time * space * rows_sent * space * rows_examined * sep
 
 local use_db        = l.P"use " * line
 
-local last_insert   = l.P"last_insert_id=" * l.digit^1 * ","
-local insert        = l.P"insert_id=" * l.digit^1 * ","
-local timestamp     = l.P"timestamp=" * l.Cg((l.digit^1 / "%0000000000"), "Timestamp")
+local last_insert   = l.P"last_insert_id=" * l.Cg(integer, "Last_insert") * ","
+local insert        = l.P"insert_id=" * l.Cg(integer, "Insert_id") * ","
+local timestamp     = l.P"timestamp=" * l.Cg(integer / "%0000000000", "Timestamp")
```
Comment (trink) [inline modules/mysql.lua]: This now produces a string instead of a number

**Comment by trink** [comment]
I am at a work week this week hopefully I can give it a full review early next week, thanks.

File: modules/mysql.lua (hunk 1 above)
Comment (robison) [inline modules/mysql.lua]: Which part produces a string? In testing previously, I've found that the timestamp field was previously returning a string. I'll make the change to convert that to a number after the zero-padding for what I'm guessing was a form of time_to_ns conversion.

Comment (robison) [inline modules/mysql.lua]: e.g.: http://lpeg.trink.com/share/17134877670671110795

I'll change it over to something along these lines:
http://lpeg.trink.com/share/15279011581007551466

Comment (trink) [inline modules/mysql.lua]: Actually you can remove the change to the line all together and leave the original.

File: modules/mysql.lua [hunk 2]
```diff
@@ -6,35 +6,44 @@
...
+local unreserved    = l.alnum + l.S"-._~"
+local sub_delims    = l.S"!$&'()*+,;="
+
 local space         = l.space^1
 local sep           = l.P"\n"
 local sql_end       = l.P";" * (l.P"\n" + -1)
 local line          = (l.P(1) - sep)^0 * sep
-local float         = l.digit^1 * "." * l.digit^1
+local float         = l.digit^1 * "." * l.digit^1 / tonumber
+local integer       = l.P"-"^-1 * l.digit^1 / tonumber
```
Comment (trink) [inline modules/mysql.lua]: I didn't see any negative counts in the spec have you seen them in the actual output?

File: modules/mysql.lua [hunk 3]
```diff
@@ -6,35 +6,44 @@
...
+local host          = l.Ct(l.Cg(ip.v4, "value") * l.Cg(l.Cc"ipv4", "representation"))
+                    + l.Ct(l.Cg(ip.v6, "value") * l.Cg(l.Cc"ipv6", "representation"))
+                    + l.Ct(l.Cg((unreserved + pct_encoded + sub_delims)^1, "value") * l.Cg(l.Cc"hostname", "representation"))
```
Comment (trink) [inline modules/mysql.lua]: We now have something like this is multiple places.  We should probably pull host/hostname out into a common grammar module like ip_address.lua (not necessary for this PR)

File: modules/mysql.lua [hunk 4]
```diff
@@ -6,35 +6,44 @@
...
 
 local time          = l.P"# Time: " * line
 
 local user_legal    = (1 - l.S"[]")
 local host_legal    = (l.alnum + l.S".-")
-local user_name     = user_legal^1 * "[" * l.Cg(user_legal^1, "Username") * "]"
-local host_name     = host_legal^0 * l.space^0 * "[" * l.Cg((l.P(1) - "]")^1, "Hostname") * "]"
+local user_name     = user_legal^0 * "[" * l.Cg(user_legal^0, "Username") * "]"
+local host_name     = l.Cg(host, "Client_hostname")^0 * l.space^0 * "[" * l.Cg(host, "Client_address")^0 * "]"
```
Comment (trink) [inline modules/mysql.lua]: Changing the data schema Hostname->Client_address could break downstream consumers (the Hostname will now be the name of the parsing machine)

File: modules/mysql.lua [hunk 5]
```diff
@@ -6,35 +6,44 @@
...
+                    + l.Ct(l.Cg(ip.v6, "value") * l.Cg(l.Cc"ipv6", "representation"))
+                    + l.Ct(l.Cg((unreserved + pct_encoded + sub_delims)^1, "value") * l.Cg(l.Cc"hostname", "representation"))
 
 local time          = l.P"# Time: " * line
 
 local user_legal    = (1 - l.S"[]")
 local host_legal    = (l.alnum + l.S".-")
-local user_name     = user_legal^1 * "[" * l.Cg(user_legal^1, "Username") * "]"
-local host_name     = host_legal^0 * l.space^0 * "[" * l.Cg((l.P(1) - "]")^1, "Hostname") * "]"
+local user_name     = user_legal^0 * "[" * l.Cg(user_legal^0, "Username") * "]"
```
Comment (trink) [inline modules/mysql.lua]: I will need to double check to see if user_name can validly be empty.

File: modules/mysql.lua [hunk 6]
```diff
@@ -43,22 +52,74 @@ local sql           = l.Cg((l.P(1) - sql_end)^0 * sql_end, "Payload")
 
 -- Maria DB extensions
 local yes_no        = l.C(l.P"Yes" + "No")
-local thread_id     = l.P"# Thread_id: " * l.Cg(l.digit^1 / tonumber, "Thread_id")
-                    * l.P"  Schema: " * l.Cg(l.alnum^0, "Schema")
+local thread_id     = l.P"# Thread_id: " * l.Cg(integer, "Thread_id")
+                    * l.P"  Schema: " * l.Cg(unreserved^1, "Schema")
```
Comment (trink) [inline modules/mysql.lua]: Schema can be an empty string on the ^0 is needed

File: modules/mysql.lua [hunk 7]
```diff
@@ -43,22 +52,74 @@ local sql           = l.Cg((l.P(1) - sql_end)^0 * sql_end, "Payload")
...
 mariadb_slow_query_grammar  = l.Ct(time^0 * user * l.Cg(l.Ct(thread_id * query * full_scan^0), "Fields") * use_db^0 * set * admin^0 * sql)
+percona_slow_query_grammar  = l.Ct(time^0 * percona_user * l.Cg(l.Ct(conn_id^0 * info^0 * percona_query * memory_footprint^0 * stored_routine^0 * query_plan_info^0 * innodb_usage^0), "Fields") * use_db^0 * set * admin^0 * sql)
 
-short_slow_query_grammar        = l.Ct(l.Cg(l.Ct(query), "Fields") * use_db^0 * set * admin^0 * sql)
-mariadb_short_slow_query_grammar= l.Ct(l.Cg(l.Ct(thread_id * query * full_scan^0), "Fields") * use_db^0 * set * admin^0 * sql)
+short_slow_query_grammar         = l.Ct(l.Cg(l.Ct(query), "Fields") * use_db^0 * set * admin^0 * sql)
+mariadb_short_slow_query_grammar = l.Ct(l.Cg(l.Ct(thread_id * query * full_scan^0), "Fields") * use_db^0 * set * admin^0 * sql)
 
 return M
+
```
Comment (trink) [inline modules/mysql.lua]: The extra blank line can be remove, the other reformatting is fine.  The grammar looks reasonable (without looking at the Percona source) but I am betting Schema can also be an empty string in their output too.

File: modules/mysql.lua (hunk 2 above)
Comment (robison) [inline modules/mysql.lua]: I believe that I've seen them previously, but I can't seem to find any concrete examples to share at the moment.

File: modules/mysql.lua (hunk 3 above)
Comment (robison) [inline modules/mysql.lua]: Should this remain for now?

File: modules/mysql.lua (hunk 4 above)
Comment (robison) [inline modules/mysql.lua]: Ah, good point. I'll change this back to Hostname

File: modules/mysql.lua (hunk 7 above)
Comment (robison) [inline modules/mysql.lua]: Oversight on my part.

**Comment by robison** [comment]
Updated this with the suggested inline changes in a squashed commit.

File: modules/mysql.lua (hunk 1 above)
Comment (robison) [inline modules/mysql.lua]: Just saw that I missed this; changing it back to the original timestamp pattern.

Comment (trink) [inline modules/mysql.lua]: I meant the original (before this PR).  It saves the conversion from string to number, to a string, and then back to number.

File: src/test/lua/lpeg_mysql.lua [hunk 8]
```diff
@@ -139,45 +224,56 @@ end
 local function header()
     local t = mysql.slow_query_grammar:match(slow_query_log[1])
     if not t then return error("no match") end
-    if t.Hostname ~= "127.0.0.1" then return error("Hostname:" .. t.Hostname) end
+    if t.Hostname ~= "127.0.0.1" then return error("Address: " .. t.Hostname) end
```
Comment (trink) [inline src/test/lua/lpeg_mysql.lua]: Why change the error message(s)?

Comment (robison) [inline src/test/lua/lpeg_mysql.lua]: Unintentional leftover from my changing to `Client_address`. I'll fix this up.

File: modules/mysql.lua (hunk 1 above)
Comment (robison) [inline modules/mysql.lua]: Fixed this. I had a fundamental misunderstanding about what was going on here.

**Comment by trink** [comment]
Thanks, for the contribution

**Comment by robison** [comment]
Thanks for the guidance.

**Pull request merged**